Add new IcfgBuilder

Heizmann opened the pull request:
Commits for switching from RCFGBuilder to IcfgBuilder are missing in this branch. But Frank convinced me to already start a pull request.

NiklasKult committed:
wip IcfgBuilder Plugin

NiklasKult committed:
wip 2 IcfgBuilder Plugin

NiklasKult committed:
moving of classes into old RCFGBuilder & invariants

NiklasKult committed:
Backtranslation wip

NiklasKult committed:
Backtranslation

NiklasKult committed:
cleanup

Heizmann committed:
Rename license header

NiklasKult committed:
comment and delete useless file

NiklasKult committed:
add author info

schuessf committed:
Fix for new ICFG in WitnessPrinter

schuessf committed:
Bugfix: Do not just reuse ateBuilder

Otherwise the StepInfo is reset, since mutable sets are used.

NiklasKult committed:
fix label last statement in procedure

schuessf committed:
Add IcfgBuilder to feature.xml

NiklasKult committed:
fix Label last Statement of While/If

Heizmann committed:
Omit UnstructureCode in BoogiePreprocessor

Heizmann committed:
Fix plugin name

Heizmann committed:
Require plugin instead of require package

Heizmann committed:
Add icfgbuilder to UltimateRegressionTest

danieldietsch committed:
add icfgbuilder as valid generator for abstract interpretation plugin

Heizmann committed:
Adapt to Dominik's move of AtomicBlockInfo

Heizmann committed:
CC fixes

Heizmann committed:
Simplify code: more explicit case distinction

Heizmann committed:
Bugfix: We need the target location, new StatementSequence is startet when we build the new block

Heizmann committed:
Inline method

Heizmann committed:
Bugfix BooleanLiteral

We need that BooleanLiteral have a type.
We now ensure this by using the ExpressionFactory#createBooleanLiteral

maul-esel committed:
move CfgBuilder.LargeBlockEncoding to own file

maul-esel committed:
avoid duplicating CFG LargeBlockEncoding

Heizmann committed:
Fix backtranslation of IfStatement and WhileStatement

Heizmann committed:
Revert "Inline method"

This reverts commit f13a388d7d265aae78824f92e166720277bfd467.

Heizmann committed:
Revert "Bugfix: We need the target location, new StatementSequence is startet when we build the new block"

This reverts commit ba104cfb44c41939202257e9bf9c830a465a75d0.

Heizmann committed:
Bugfix atomic blocks in IcfgBuilder

Bugfix for cases where atomic blocks end with an IfStatement.

Heizmann committed:
Explicit case distinction, cast to StatementSequence

Heizmann committed:
Adapt solver check to IcfgBuilder

Heizmann committed:
Add IcfgBuilder to BuchiProgramProduct

Heizmann committed:
Refactor private method of IcfgBuilder

* Rename identifiers
* Add documentation
* Make method void (returned object is an argument of method)

Heizmann committed:
Refactor private method of IcfgBuilder

* Rename identifiers
* Add documentation
* Make method void (returned object is an argument of method)

Heizmann committed:
Refactor IcfgBuilder

* extract method for building IcfgLocations
* remove method that became obsolete

Heizmann committed:
Bugfix and refactoring IcfgBuilder

* Bugfix: Do not merge loop entry and location after invariant.
* Rename identifiers

Heizmann committed:
Add backtranslation for free invariants

Heizmann committed:
Bugfix: Support for backtranslation of loop invariants

Heizmann committed:
Update errorpath

Heizmann committed:
Refactoring IcfgBuilder

Do not copy annotations directly from original statements to edges but only to
auxiliary statements. (Because annotations are copied from all
statements to edges anyway.)

Heizmann committed:
Refactoring IcfgBuilder

Unify code for prepending statements

Heizmann committed:
Remove origin

Heizmann committed:
Add documentation and assert

Heizmann committed:
Add auxiliary edges for labels

Heizmann committed:
Refactoring IcfgBacktranslator

Use assumption that mapping returns singleton except for calls.

Heizmann committed:
Omit labels in backtranslation

Heizmann committed:
Bugfix backtranslation

Heizmann committed:
Remove XnfConversionTechnique

schuessf commented on trunk/examples/settings/automizer/ForwardPredicates_IcfgBuilder.epf:
```diff
@@ -0,0 +1,27 @@
+#Thu Jan 30 13:20:28 CET 2014
```
These settings can be removed, once `IcfgBuilder` is the default.

schuessf commented on trunk/examples/toolchains/AutomizerBpl_IcfgBuilder.xml:
```diff
@@ -0,0 +1,10 @@
+<rundefinition>
```
This toolchain can be removed, once `IcfgBuilder` is the default.

schuessf commented on trunk/examples/Backtranslation/regression/bpl/Pointer.errorpath:
```diff
@@ -1,38 +1,40 @@
-       VAL        [#NULL!base=0, #NULL!offset=0]
-[L31]             assert false;
-       VAL        [#NULL!base=0, #NULL!offset=0]
+[L49]  CALL       call ULTIMATE.init();
```
What has changed in the backtranslation that is relevant for this test?

Heizmann replied:
Could you fix this?

schuessf replied:
I was just wondering why you changed already this path and not the others.

schuessf commented on trunk/source/IcfgBuilder/src/de/uni_freiburg/informatik/ultimate/plugins/generator/icfgbuilder/preferences/IcfgPreferenceInitializer.java:
```diff
@@ -0,0 +1,157 @@
+import de.uni_freiburg.informatik.ultimate.lib.smtlibutils.solverbuilder.SolverBuilder.SolverMode;
+import de.uni_freiburg.informatik.ultimate.plugins.generator.icfgbuilder.Activator;
+
+public class IcfgPreferenceInitializer extends UltimatePreferenceInitializer {
```
Please make sure that these settings are actually the same as for `RCFGBuilder` (this was at least not the case in the past for `CodeBlockSize`).

Heizmann replied:
Could you fix this?

schuessf replied:
I am not sure, if there is a problem in the current version. I just remember that there was some issue in the past. So this comment is more of a reminder 😉

schuessf replied:
I just checked, the settings are synced right now.

schuessf commented on trunk/source/IcfgBuilder/src/de/uni_freiburg/informatik/ultimate/plugins/generator/icfgbuilder/WeakestPrecondition.java:
```diff
@@ -0,0 +1,188 @@
+import de.uni_freiburg.informatik.ultimate.boogie.ast.VarList;
+import de.uni_freiburg.informatik.ultimate.core.model.models.ModelUtils;
+
+public class WeakestPrecondition extends BoogieTransformer {
```
I suppose this class (along with some others) was simply copied from `RCFGBuilder` without any change? I think we should rather have another plugin `Library-CFG` with such classes, s.t. `RCFGBuilder` could be safely deleted.

Heizmann replied:
Plan is to move the classes to from `RCFGBuilder` to `IcfgBuilder` if we want to get rid of the `RCFGBuilder`.

schuessf replied:
I see, this class is really just used in `IcfgBuilder` / `RCFGBuilder`, so it makes sense to keep it there. But there are some other classes (e.g. `TransFormulaAdder`) that are also used in other Plugins. So it would make sense to move them to some library, also to reduce the dependencies to `IcfgBuilder` / `RCFGBuilder`.

schuessf replied:
Okay, for now we don't create such a library, as this PR should be merged asap

schuessf commented on trunk/examples/settings/automizer/ForwardPredicates_IcfgBuilder.epf:
```diff
@@ -0,0 +1,27 @@
+#Tue Mar 19 19:15:08 CET 2024
+file_export_version=3.0
+\!/instance/de.uni_freiburg.informatik.ultimate.boogie.preprocessor=
+/instance/de.uni_freiburg.informatik.ultimate.boogie.preprocessor/Replace\ while\ statements\ and\ if-then-else\ statements=false
```
This should be our new default now, maybe you can simply change it in `BoogiePreprocessor`.

schuessf commented on trunk/source/RCFGBuilder/src/de/uni_freiburg/informatik/ultimate/plugins/generator/rcfgbuilder/cfg/LargeBlockEncoding.java:
Why was this class added to `RCFGBuilder`?

maul-esel replied:
I extracted this class from `CfgBuilder` (where it used to be a nested class) so it can be used by both implementations.

schuessf reviewed (commented):
Thanks for your work! I did not have a detailed look yet, but here are already some more general comments.

maul-esel committed:
make sure nodatarace-LBE is only used for concurrent programs

The setting "Only consider context switches at boundaries of atomic blocks"
(aka the nodatarace-LBE) is only meant to affect concurrent programs. It
performs a kind of large-block encoding (LBE) that is usually not desirable
for sequential programs.

schuessf committed:
Use IcfgBuilder in all settings and toolchains

schuessf committed:
CC fixes

schuessf committed:
Sync preferences

schuessf committed:
Remove IcfgBuilder settings and toolchains
IcfgBuilder is already the default anyway.

schuessf committed:
Make class public to use in IcfgBuilder

schuessf committed:
Sync CfgBuilder classes

schuessf commented:
I just synced the `CfgBuilder` classes from `RCFGBuilder` and the new `IcfgBuilder`. To inspect the differences, you can take a look a the [diff](https://github.com/user-attachments/files/17678265/cfgbuilder.patch).

schuessf commented on trunk/source/IcfgBuilder/src/de/uni_freiburg/informatik/ultimate/plugins/generator/icfgbuilder/cfg/CfgBuilder.java:
```diff
@@ -0,0 +1,1627 @@
+ * @author heizmann@informatik.uni-freiburg.de
+ */
+
+// TODO How to give every location the right line number
```
Is this TODO fixed?

schuessf commented on trunk/source/IcfgBuilder/src/de/uni_freiburg/informatik/ultimate/plugins/generator/icfgbuilder/cfg/CfgBuilder.java:
```diff
@@ -0,0 +1,1627 @@
+		}
+
+		private IIcfgElement buildAssuAssiHavoc(final IIcfgElement currentElement, Statement st) {
+			// FIXME Matthias 2024-09-22: This code is extremely problematic because it
```
@Heizmann Are you planning to change this? Why isn't this necessary in the old `RCFGBuilder`?

schuessf commented:
@Heizmann I just ran the nightly build. Unfortunately, currently not only the backtranslation tests fail (which is expected and could be quite easily fixed), but also [generic regression tests](https://jenkins.sopranium.de/job/Ultimate/job/Ultimate%20Nightly/job/PR-690/1/testReport/de.uni_freiburg.informatik.ultimate.regressiontest.generic/).
This includes some LTL tests (e.g. [here](https://jenkins.sopranium.de/job/Ultimate/job/Ultimate%20Nightly/job/PR-690/1/testReport/de.uni_freiburg.informatik.ultimate.regressiontest.generic/LTLRegressionTestSuite/I_LTL_regression_c_stepwise_baseline_coolant_basis_4_neg_c_S_LTL_regression_c_LTLAutomizerC_epf_T_LTL_regression_c_LTLAutomizerC_xml/)), where we don't find any specification, some referee tests (e.g. [here](https://jenkins.sopranium.de/job/Ultimate/job/Ultimate%20Nightly/job/PR-690/1/testReport/de.uni_freiburg.informatik.ultimate.regressiontest.generic/RegressionTestSuite/I_programs_toy_InvariantChecking_regression_bpl_OnlyRequires01_bpl_S_programs_toy_InvariantChecking_regression_bpl_RefereeBpl_epf_T_programs_toy_InvariantChecking_regression_bpl_RefereeBpl_xml/)), where we claim that there is a cycle now and some concurrency tests (e.g. [here](https://jenkins.sopranium.de/job/Ultimate/job/Ultimate%20Nightly/job/PR-690/1/testReport/de.uni_freiburg.informatik.ultimate.regressiontest.generic/RegressionTestSuite/I_concurrent_bpl_regression_concurrent_increment_bpl_S_concurrent_bpl_regression_ReachSafety_32bit_GemCutter_epf_T_concurrent_bpl_regression_ReachSafety_xml/)) that now have a timeout, maybe because of the `assume true` statements that are not removed (see [above](https://github.com/ultimate-pa/ultimate/pull/690#issuecomment-2455268452))

schuessf commented on trunk/source/IcfgBuilder/src/de/uni_freiburg/informatik/ultimate/plugins/generator/icfgbuilder/cfg/CfgBuilder.java:
```diff
@@ -0,0 +1,1627 @@
+				firstStatement =
+						new ReturnStatement(mBoogieDeclarations.getProcImplementation().get(procName).getLocation());
+			} else {
+				firstStatement = body.getBlock()[0];
```
In the `RCFGBuilder` `firstStatement` is always set to `body.getBlock()[0]`, why is this different here?

NiklasKult replied:
For the <code class="notranslate">RCFGBuilder</code> <code class="notranslate">body.getBlock()</code> could never be empty, because the <code class="notranslate">BoogiePreprocessor</code> would insert a return statement in that case. With the setting that disables the <code class="notranslate">UnstructureCode</code> class from the <code class="notranslate">BoogiePreprocessor</code> <code class="notranslate">body.getBlock()</code> can now be empty.

schuessf replied:
But does this belong to the `CfgBuilder`? Wouldn't it be better to move it into its own Preprocessor that always runs independant of `UnstructureCode`?

schuessf replied:
Or do we even need this statement here? Is it possible to create a location without a statement?

schuessf commented on trunk/source/IcfgBuilder/src/de/uni_freiburg/informatik/ultimate/plugins/generator/icfgbuilder/cfg/CfgBuilder.java:
```diff
@@ -0,0 +1,1627 @@
+				mIcfgBacktranslator.putAux(st, new BoogieASTNode[] { spec });
+				mCurrent = buildAssuAssiHavoc((IIcfgElement) mCurrent, st);
+			}
+			final BoogieIcfgLocation finalNode = mIcfg.mFinalNode.get(mCurrentProcedureName);
```
In the `RCFGBuilder` the exit nodes are used, while here the final nodes are used. Is this change intended and what are the implications?

NiklasKult replied:
The RCFGBuilder builds from the top, while the IcfgBuilder builds from the bottom. Because of that the ensures is built from the exit node instead of the final node.
The resulting CFG should be identical here.

schuessf committed:
Move replacement of array assignment from UnstructureCode to new observer
The new IcfgBuilder does not use UnstructureCode anymore. To avoid reimplementation there, this was moved to the new oberserver ReplaceArrayAssignments.

schuessf committed:
Add removal of assume true statements
This does not seem as useful as in the RCFGBuilder, as most of such assume true statements where introduced in UnstructureCode, but it does not harm to keep it.

schuessf committed:
Fix sonar issues, remove unused code

schuessf committed:
Move side-effect outside of assertion

schuessf commented:
I just looked into the open backtranslation issues, the created CFG for loops (especially when translated from C) looks slightly different.
Consider the following C program
```c
while (x > 0) x--
```
This is roughly translated to the following Boogie program
```c
while (true) {
  label: if (!(x > 0)) {
    break;
  }
  x := x - 1;
}
```
In the old CFG, we only had only two states and three edges (with `SequenceOfStatements`):
* a loop in the initial location with the body and the negated if-condition (`x > 0`)
* an edge from the initial location to the exit location with the if-condition (`!(x > 0)`)
* an edge from the initial location to the exit location with the negated while-condition (`false`)
![grafik](https://github.com/user-attachments/assets/6706de65-0d94-4bc5-b1a8-cfbbd1f6ce47)


With the same settings, the new CFG has 5 locations and the following 6 edges:
* one with the while condition (`true`)
* an edge from the initial location to the exit location with the negated while-condition (`false`)
* another edge with `true` after the label
* an edge with the negated if-condition (`x > 0`)
* an edge to the exit location with the if-condition (`!(x > 0)`)
* an edge with the body back to the initial location
![grafik](https://github.com/user-attachments/assets/cbcc52d9-e800-4689-a887-1e3e3c8e592a)

This means that in the old CFG we had a stronger block encoding for `SequenceOfStatements`. With other settings the CFG should be more similar. I am not sure if this behaviour is desired.

Heizmann commented:
Some info (not enough time to think about your post): The new IcfgBuilder alredy makes sure that labels get their own IcfgLocation.

NiklasKult committed:
Remove if end location if possible

schuessf committed:
Adapt to e53bbfe

NiklasKult committed:
fix reachable but not registered program points

schuessf committed:
Fix missing annotation

schuessf committed:
Minor simplifications

bahnwaerter commented on trunk/source/IcfgBuilder/.classpath:
```diff
@@ -0,0 +1,8 @@
+	<classpathentry kind="con" path="org.eclipse.jdt.launching.JRE_CONTAINER/org.eclipse.jdt.internal.debug.ui.launcher.StandardVMType/JavaSE-21"/>
+	<classpathentry exported="true" kind="con" path="org.eclipse.pde.core.requiredPlugins"/>
+	<classpathentry kind="src" path="src"/>
+	<classpathentry combineaccessrules="false" kind="src" path="/BoogiePrinter"/>
```
Why is this access rule necessary? Does a utility need to be accessed internally here?
If so, then the API and implementation of the required software component should be restructured to avoid such hacks and instead access public interfaces cleanly via the `MANIFEST.MF`.

schuessf replied:
I don't know, this is the same as in `RCFGBuilder`, so I will just keep this for now, we can change this later (after SV-COMP 😉)

bahnwaerter commented on trunk/source/IcfgBuilder/LicenseTemplate.md:
```diff
@@ -0,0 +1,23 @@
+Copyright (C) @{daterange} @{author:r}
+
+This file is part of the ULTIMATE RCFGBuilder plug-in.
```
Shouldn't the objective be called ICFGBuilder instead of RCFGBuilder in the whole license template file?

bahnwaerter commented on trunk/source/IcfgBuilder/src/de/uni_freiburg/informatik/ultimate/plugins/generator/icfgbuilder/IcfgBacktranslator.java:
```diff
@@ -0,0 +1,471 @@
+			translateCFG(final IBacktranslatedCFG<IcfgLocation, IIcfgTransition<IcfgLocation>> cfg) {
+		final IBacktranslatedCFG<String, BoogieASTNode> translatedCfg =
+				translateCFG(cfg, (a, b, c) -> translateCFGEdge(a, (IIcfgTransition<IcfgLocation>) b, c));
+		// mLogger.info(getClass().getSimpleName());
+		// printHondas(cfg, mLogger::info);
+		// printCFG(cfg, mLogger::info);
+		// mLogger.info("######## END "+getClass().getSimpleName());
```
Remove unused debugging code or implement it correctly when debug mode is enabled.

bahnwaerter reviewed (commented):
The changes look good so far. The branch has already been correctly adapted to our Java 21 and Eclipse migration and can therefore be merged without any issues.

However, there are still a few small things that need to be improved. For example, in the comments and the license agreement in the headers of ICFG source code files, the word _RCFGBuilder_ should be replaced consistently with _ICFGBuilder_.

schuessf committed:
Revert "Omit UnstructureCode in BoogiePreprocessor"

This reverts commit 506cd1a4e037544fda09abcb14035c6d0ca4a451.

schuessf committed:
Revert "Adapt solver check to IcfgBuilder"

This reverts commit 0ec76e32a647c0304ca771f7a89e1431d7ffb3f8.

schuessf committed:
Revert "Use IcfgBuilder in all settings and toolchains"

This reverts commit 1b2b065a5937a04b49a9486162f37a122812094b.

schuessf committed:
Add IcfcBuilder settings to Automizer reachability settings

schuessf committed:
Add Automizer toolchain with IcfgBuilder

schuessf committed:
Use new IcfgBuilder for witness validation

schuessf committed:
Fix LicenseTemplate.md

schuessf committed:
Remove outcommented code

schuessf committed:
Revert "Update errorpath"

This reverts commit 82df167d589fd1c608626168ed2a445d57c48c30.

schuessf committed:
Use IcfgBuilder in WitnessRegressionTestSuite
to check that validation in SV-COMP actually works

maul-esel commented on trunk/source/AbstractInterpretationV2/src/de/uni_freiburg/informatik/ultimate/plugins/analysis/abstractinterpretationv2/AbstractInterpretation.java:
```diff
@@ -77,10 +77,12 @@ public void setInputDefinition(final ModelType graphType) {
 			mObserver.add(new AbstractInterpretationRcfgObserver(mServices));
 			break;
 		default:
+			mLogger.info("Ignoring creator " + creator);
```
Nitpick: Make this a warning

maul-esel commented on trunk/source/IcfgBuilder/src/de/uni_freiburg/informatik/ultimate/plugins/generator/icfgbuilder/IcfgBacktranslator.java:
```diff
@@ -0,0 +1,467 @@
+	@Override
+	public IProgramExecution<BoogieASTNode, Expression>
+
+			translateProgramExecution(final IProgramExecution<IIcfgTransition<IcfgLocation>, Term> programExecution) {
```
nitpick: Fix the formatting

maul-esel commented on trunk/source/IcfgBuilder/src/de/uni_freiburg/informatik/ultimate/plugins/generator/icfgbuilder/IcfgBacktranslator.java:
```diff
@@ -0,0 +1,467 @@
+
+			translateProgramExecution(final IProgramExecution<IIcfgTransition<IcfgLocation>, Term> programExecution) {
+		if (!(programExecution instanceof IcfgProgramExecution)) {
+			throw new IllegalArgumentException();
```
Nitpick: add an error message

maul-esel commented on trunk/source/IcfgBuilder/src/de/uni_freiburg/informatik/ultimate/plugins/generator/icfgbuilder/cfg/CfgBuilder.java:
```diff
@@ -0,0 +1,1617 @@
+						.setUseExternalSolver(true, commandExternalSolver, logicForExternalSolver)
+						.setSolverMode(solverMode).setAdditionalOptions(additionalSmtOptions);
+
+		return SolverBuilder.buildAndInitializeSolver(services, solverSettings, "CfgBuilderScript");
```
In the future, we could wrap these settings in a suitable class rather than accessing the `IPreferenceProvider` directly here.
This should not be an obstacle for the merge now though.

maul-esel commented on trunk/source/IcfgBuilder/src/de/uni_freiburg/informatik/ultimate/plugins/generator/icfgbuilder/cfg/CfgBuilder.java:
```diff
@@ -0,0 +1,1617 @@
+					&& endLoc.getIncomingEdges().size() == 1
+					&& endLoc.getOutgoingEdges().size() == 1
+					&& !mConditionalStarts.contains(endLoc)
+					&& !mLabel2LocNodes.containsValue(endLoc)) {
```
A complex condition like that could be encapsulated in a method with an intuitive name.

maul-esel commented on trunk/source/RCFGBuilder/src/de/uni_freiburg/informatik/ultimate/plugins/generator/rcfgbuilder/cfg/BoogieIcfgContainer.java:
```diff
@@ -89,7 +89,7 @@ public class BoogieIcfgContainer extends ModernAnnotations implements IIcfg<Boog
 	 *
 	 */
-	final Map<String, BoogieIcfgLocation> mFinalNode;
+	public final Map<String, BoogieIcfgLocation> mFinalNode;
```
Is it necessary to make this public?

schuessf replied:
I think so, as this is also used by the IcfgBuilder plugin. Of course, it would be more optimal to move it to some library, but we can do this, if we remove RCFGBuilder in the future.

maul-esel reviewed (approved):
Thanks @NiklasKult, @Heizmann and @schuessf for working on this!

I had a look at the changes and did not find any major issues. Due to the time constraints, I was not able to review in-depth, but since others have done so, and our latest evaluations have shown no major issues (only a slight increase in the number of locations) I have no objection to merging :+1:

maul-esel commented:
I also had a quick look at the RCFGBuilder/ICFGBuilder diff by @schuessf and did not find any issues there either.

schuessf committed:
CC fixes

schuessf commented:
I will try to address these minor issues and merge this later.

schuessf committed:
Mark message as warning

schuessf committed:
Add error message

schuessf committed:
CC fixes

schuessf committed:
Extract condition to method
